### ISSUE TITLE
kernel dying calling result of `jit(grad(f))`, but not `jit(f)` or `grad(f)`

### ISSUE THREAD

**Issue by DanPuzzuoli** [google/jax#4687]
Hi,

I'm running into a situation with jax in which I have a function `f` from an array of floats to a float, in which:
- `grad(f)` works (i.e. I can call the resulting function and it evaluates),
- `jit(f)` works (i.e. I can call the resulting function and it evaluates), but
- `jit(grad(f))` causes the kernel to die once I call the resulting function

As there is no error message I have no idea how to debug this. The function is very complicated; I'm going to try breaking it down into constituent pieces to see if any of them fail in this way, but figured I'd post here in the meantime in case there are any hints. In case it is relevant, part of this function involves a call to `jax.experimental.ode.ode_int` (though I haven't had any issue with this in the past). 

Any thoughts?

Thanks!

**Comment by jekbradbury** [comment]
When you say the kernel dies, I assume you mean a Colab or Jupyter kernel? If you can get the kernel logs, they'll likely show a more detailed error; alternatively, if you post a reproducible example (even if it's long) we can get that error ourselves.

**Comment by DanPuzzuoli** [comment]
Hi,

Thanks for the response. Yes sorry I forgot to specify, I'm running code in a jupyter notebook.

I've written a much smaller example below. The purpose is to simulate a DE of the form 
```
y'(t) = G(c, t)y(t),
```
where `G` is a matrix with some parameters `c` (given as a 1d array), and we want to compute some objective function of the solution `y(T)`, along with gradients in the entries of `c`. In the code below (as well as in my original code mentioned in the first comment), `G(c,t)` is piecewise constant, with `c` essentially defining "sample values". To compute `G(c,t)`, the correct "sample" needs to be looked up in `c`, which is determined by an additional `dt` parameter giving the sample width.

```
import jax.numpy as np
from jax.experimental.ode import odeint
from jax import jit, value_and_grad

dim = 2
X = np.diag(np.ones(dim - 1, dtype=complex), 1) + np.diag(np.ones(dim - 1, dtype=complex), -1)

# define G(c,t)
def G(c, dt, t):
    """c are samples in a piecewise constant signal, with
    sample width dt
    """
    idx = np.array(t // dt).astype(int) # could this line be problematic?
    return c[idx]*(-1j * X)

# define function of solution of the DE
def func(A):
    ip = (X * A).sum()
    return np.real(ip * ip.conj()/(dim**2))

# define solving and objective evaluation function
y0 = np.eye(dim, dtype=complex)
def basic_solve(G, c, dt, tol=1e-12):
    T = len(c) * dt
    
    def rhs(y, t):
        return G(c, dt, t) @ y
    
    return func(odeint(rhs, y0=y0, t=np.array([0, T]), atol=tol, rtol=tol)[-1])
```

Using this, I can run:
```
dt = 0.4
basic_solve_jit = jit(lambda c: basic_solve(G, c, dt))
basic_solve_jit(np.ones(2, dtype=float))
```
with no issue. But if I then run:

```
basic_solve_grad_jit = jit(value_and_grad(lambda c: basic_solve(G, c, dt)))
basic_solve_grad_jit(np.ones(2, dtype=float))[0].block_until_ready()
```
the jupyter notebook kernel dies.

The Kernel logs give:
```
2020-11-03 09:29:30.614204: F external/org_tensorflow/tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:1542] Check failed: ShapeUtil::Compatible(pad->shape(), inferred_return_shape) return shape is set to: f64[2] but is inferred to be: f64[3]
```

Not sure how these shapes are being inferred. Above I'm passing in `c` with length `2`. The error persists up until `c` with length `4`, but if `c` is length `5` the code runs fine.

Thanks again!

**Comment by DanPuzzuoli** [comment]
Closing this, I haven't fully solved this, but in retrospect it may have been due to a rounding error which eventually propagated into an index choice in an array.

**Comment by hawkinsp** [comment]
There is certainly still a bug for us. JAX shouldn't crash the Python process.

**Comment by DanPuzzuoli** [comment]
Ah makes sense, apologies for closing it!

**Comment by majnemer** [comment]
Should be fixed in XLA with https://github.com/tensorflow/tensorflow/commit/712be8cb4b002bbc1b067eca79de8c01180e864c.

**Comment by hawkinsp** [comment]
#5676 will incorporate this fix into jaxlib, although you'll have to rebuild jaxlib from source or wait for us to make a new jaxlib release. Hope that helps!
